### PR TITLE
Deprecation: notification and feature flag for `build.image` config

### PR DESCRIPTION
**Pull request by humitos**
Define a weekly task to communicate our users about the deprecation of `build.image` (or not specifying `build.os`) using the deprecation plan we used for the configuration file v2 as well.

- 3 brownout days
- final removal date on October 16th
- weekly onsite/email notification on Wednesday at 11:15 CEST (around ~16k projects affected)
- allow to opt-out from these emails
- feature flag for brownout days
- build detail's page notification

Related:
* https://github.com/readthedocs/meta/discussions/48
* https://github.com/readthedocs/readthedocs.org/pull/10354
* https://github.com/readthedocs/readthedocs.org/issues/10587
* https://github.com/readthedocs/readthedocs-ops/pull/1383
* https://ethicalads.metabaseapp.com/question/473-projects-not-specifying-build-image-nor-build-os-config-file?days=30

Closes:
* https://github.com/readthedocs/readthedocs-docker-images/issues/193

### CODE REVIEW

File: readthedocs/core/models.py [hunk 1]
```diff
@@ -51,6 +51,13 @@ class UserProfile(TimeStampedModel):
         default=False,
         null=True,
     )
+    # NOTE: this is a temporary field that we can remove after October 16, 2023
+    # See https://blog.readthedocs.com/build-image-config-deprecated/
```
Comment (ericholscher) [inline readthedocs/core/models.py:55]: The blog file is: use-build-os-config.rst

File: readthedocs/projects/tasks/utils.py [hunk 2]
```diff
@@ -323,6 +323,160 @@ def deprecated_config_file_used_notification():
     )
 
 
+class DeprecatedBuildImageSiteNotification(SiteNotification):
+    failure_message = _(
+        'Your project(s) "{{ project_slugs }}" are using the deprecated "build.image" '
```
Comment (ericholscher) [inline readthedocs/projects/tasks/utils.py:328]: Would be nice to link to these projects, I've found it annoying not to be able to click the link to the project, but not a huge deal.

File: readthedocs/projects/tasks/utils.py [hunk 3]
```diff
@@ -323,6 +323,160 @@ def deprecated_config_file_used_notification():
...
+
+        # Only check for the default version because if the project is using tags
+        # they won't be able to update those and we will send them emails forever.
+        # We can update this query if we consider later.
+        version = (
+            project.versions.filter(slug=project.default_version).only("id").first()
+        )
+        if version:
+            # Use a fixed date here to avoid changing the date on each run
+            years_ago = timezone.datetime(2020, 8, 1)
```
Comment (ericholscher) [inline readthedocs/projects/tasks/utils.py:397]: This is in the last 3 years, which feels like a long time? Is this what we want?

File: readthedocs/templates/projects/notifications/deprecated_build_image_used_email.txt [hunk 4]
```diff
@@ -0,0 +1,31 @@
+{% extends "core/email/common.txt" %}
+{% block content %}
+The Read the Docs build system is deprecating "build.image" config key on ".readthedocs.yaml" starting on October 16, 2023.
+We are sending weekly notifications about this issue to all impacted users,
+as well as temporary build failures (brownouts) as the date approaches for those who haven't migrated their projects.
+
```
Comment (ericholscher) [inline readthedocs/templates/projects/notifications/deprecated_build_image_used_email.txt:6]:
```suggestion
The "build.image" config key on ".readthedocs.yaml" has been deprecated, and will be removed on October 16, 2023.
We are sending weekly notifications about this issue to all impacted users,
as well as temporary build failures (brownouts) as the date approaches for those who haven't migrated their projects.

```

File: readthedocs/templates/projects/notifications/deprecated_build_image_used_email.txt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+The timeline for this deprecation is as follows:
+
+* Monday, August 28, 2023: Do the first brownout (temporarily enforce this deprecation) for 12 hours: 00:01 PST to 11:59 PST (noon)
+* Monday, September 18, 2023: Do a second brownout (temporarily enforce this deprecation) for 24 hours: 00:01 PST to 23:59 PST (midnight)
+* Monday, October 2, 2023: Do a third and final brownout (temporarily enforce this deprecation) for 48 hours: 00:01 PST to October 3, 2023 23:59 PST (midnight)
+* Monday, October 16, 2023: Fully remove support for building documentation using "build.image" on the configuration file
+
+We have identified that the following projects which you maintain, and were built in the last year, are impacted by this deprecation:
```
Comment (ericholscher) [inline readthedocs/templates/projects/notifications/deprecated_build_image_used_email.txt:14]: This says last year, so the above date should be 2022 ^^

**Review by ericholscher — COMMENTED** [review]
This looks good to me, and similar to our other approach. A few questions, but I think it's close.

File: readthedocs/doc_builder/exceptions.py [hunk 6]
```diff
@@ -66,6 +66,15 @@ class BuildUserError(BuildBaseException):
         "Add a configuration file to your project to make it build successfully. "
         "Read more at https://docs.readthedocs.io/en/stable/config-file/v2.html"
     )
+    BUILD_IMAGE_CONFIG_KEY_DEPRECATED = gettext_noop(
+        'The configuration key "build.image" is deprecated. '
+        'Please, use "build.os" instead to make it build successfully. '
```
Comment (agjohnson) [inline readthedocs/doc_builder/exceptions.py:71]: A couple copy changes:

- We can be more authoritative without "please"
- The "it" is ambiguous
- Avoid "successfully" as a successful build is up to the maintainer

```suggestion
        'Use "build.os" instead to continue building your project. '
```

Comment (humitos) [reply readthedocs/doc_builder/exceptions.py:71]: Loved your suggestion with the explanation of each word 💯

File: readthedocs/projects/tasks/utils.py (hunk 2 above)
Comment (humitos) [reply readthedocs/projects/tasks/utils.py:328]: Our notification system doesn't support links in this way. It has a JS that if you click on a link from a notification it marks it as read 🤷🏼 -- which is unfortunate for the use case you are suggesting. However, I agree with you that's not great 😄

File: readthedocs/projects/tasks/utils.py (hunk 3 above)
Comment (humitos) [reply readthedocs/projects/tasks/utils.py:397]: I expanded to 3 years on purpose but I'm fine if we want to reduce it to 1.

On the config file deprecation we started with no restriction about dates and sent 65k emails only once. After that we limited to 1 year.

Here, with 3 years, I found around 16k projects, which I think it's fine.

File: readthedocs/core/models.py (hunk 1 above)
Comment (humitos) [reply readthedocs/core/models.py:55]:
```suggestion
    # See https://blog.readthedocs.com/use-build-os-config/
```

File: readthedocs/templates/projects/notifications/deprecated_build_image_used_email.txt (hunk 5 above)
Comment (humitos) [reply readthedocs/templates/projects/notifications/deprecated_build_image_used_email.txt:14]:
```suggestion
We have identified that the following projects which you maintain, and were built in the last 3 years, are impacted by this deprecation:
```

File: readthedocs/projects/tasks/utils.py (hunk 3 above)
Comment (humitos) [reply readthedocs/projects/tasks/utils.py:397]:
```suggestion
            years_ago = timezone.datetime(2022, 8, 1)
```

File: readthedocs/templates/projects/notifications/deprecated_build_image_used_email.txt [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+{% endfor %}
+{% if projects.count > 15 %}
+* ... and {{ projects.count|add:"-15" }} more projects.
+{% endif %}
+
+Please use "build.os" on your configuration file to ensure that they continue building successfully and to stop receiving these notifications.
+If you want to opt-out from these emails, you can edit your preferences in your account settings, at https://readthedocs.org/accounts/edit/.
+
+For more information on how to use "build.os",
+read our blog post at https://blog.readthedocs.com/build-image-config-deprecated/
```
Comment (humitos) [inline readthedocs/templates/projects/notifications/deprecated_build_image_used_email.txt:27]:
```suggestion
read our blog post at https://blog.readthedocs.com/use-build-os-config/
```

**Comment by humitos** [comment]
I applied all the suggestions. I will update the HTML email and we will be ready to approve and merge, I guess.

File: readthedocs/projects/tasks/utils.py (hunk 2 above)
Comment (ericholscher) [reply readthedocs/projects/tasks/utils.py:328]: Something for the notification refactor 🤔

**Review by ericholscher — APPROVED** [review]
Looks good to me. I'm fine emailing more folks at first, then dropping it down, we should just make sure the email content matches the actual timeframe used 👍

**Pull request merged**